providers/rpc/rpc.go

```diff
@@ -0,0 +1,351 @@
+package rpc
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"hash"
+	"io"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/bmc-toolbox/bmclib/v2/providers"
+	"github.com/go-logr/logr"
+	"github.com/jacobweinstock/registrar"
+)
+
+const (
+	// ProviderName for the RPC implementation.
+	ProviderName = "rpc"
+	// ProviderProtocol for the rpc implementation.
+	ProviderProtocol = "http"
+
+	// defaults
+	timestampHeader = "X-BMCLIB-Timestamp"
+	signatureHeader = "X-BMCLIB-Signature"
+	contentType     = "application/json"
+
+	// SHA256 is the SHA256 algorithm.
+	SHA256 Algorithm = "sha256"
+	// SHA256Short is the short version of the SHA256 algorithm.
+	SHA256Short Algorithm = "256"
+	// SHA512 is the SHA512 algorithm.
+	SHA512 Algorithm = "sha512"
+	// SHA512Short is the short version of the SHA512 algorithm.
+	SHA512Short Algorithm = "512"
+)
+
+// Features implemented by the RPC provider.
+var Features = registrar.Features{
+	providers.FeaturePowerSet,
+	providers.FeaturePowerState,
+	providers.FeatureBootDeviceSet,
+}
+
+// Algorithm is the type for HMAC algorithms.
+type Algorithm string
+
+// Secrets hold per algorithm slice secrets.
+// These secrets will be used to create HMAC signatures.
+type Secrets map[Algorithm][]string
+
+// Signatures hold per algorithm slice of signatures.
+type Signatures map[Algorithm][]string
+
+// Config defines the configuration for sending rpc notifications.
+type Config struct {
+	// ConsumerURL is the URL where an rpc consumer/listener is running
+	// and to which we will send and receive all notifications.
+	ConsumerURL string
+	// Host is the BMC ip address or hostname or identifier.
+	Host string
+	// Logger is the logger to use for logging.
+	Logger logr.Logger
+	// LogNotificationsDisabled determines whether responses from rpc consumer/listeners will be logged or not.
+	LogNotificationsDisabled bool
+	// Opts are the options for the rpc provider.
+	Opts Opts
+
+	// listenerURL is the URL of the rpc consumer/listener.
+	listenerURL *url.URL
+}
+
+type Opts struct {
+	// Request is the options used to create the rpc HTTP request.
+	Request RequestOpts
+	// Signature is the options used for adding an HMAC signature to an HTTP request.
+	Signature SignatureOpts
+	// HMAC is the options used to create a HMAC signature.
+	HMAC HMACOpts
+	// Experimental options.
+	Experimental Experimental
+}
+
+type RequestOpts struct {
+	// Client is the http client used for all HTTP calls.
+	Client *http.Client
```

Comment:
It feels a bit peculiar to have a client nested here as opposed to on the top `Config` object. I appreciate its related to HTTP but as far as I can tell its not really used to _construct_ a request, its used to _submit_ it.

Reply:
sure, no problem. i will move it to the top level in `Config`.